pAI expansion

Dejaku51 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
PR expanding pAI containing fixes, new features etc
- [x] https://github.com/tgstation/tgstation/pull/62916 ~~[Requires TGUI Update?]~~
- [x] https://github.com/tgstation/tgstation/pull/63160 ~~[Requires TGUI Update?]~~
- [x] https://github.com/tgstation/tgstation/pull/56404
- [x] https://github.com/tgstation/tgstation/pull/52879
- [x] ~~https://github.com/tgstation/tgstation/pull/3142~~ [Abandoned]
- [x] https://github.com/tgstation/tgstation/pull/49684
- [x] https://github.com/tgstation/tgstation/pull/62462
- [x] https://github.com/tgstation/tgstation/pull/52575
- [x] https://github.com/tgstation/tgstation/pull/57033
- [x] https://github.com/tgstation/tgstation/pull/55215
- [x] https://github.com/tgstation/tgstation/pull/61112
- [x] https://github.com/tgstation/tgstation/pull/70869 [partly]

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Added new TGUI interface for pAIs
add: pAIs have now new software: GPS
add: pAIs have now new software: Photo printer
add: Borgs now can wear drones and pAIs as hats
add: pAIs now can use binary encryption keys
add: pAIs can now select holochassis using radial menu 
tweak: pAIs can now select radio channel from encryption keys
code: Cleaned up pAIs and aicard code
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Dejaku51 committed:
Initial

Dejaku51 committed:
pAI TGUI conversion 1

Dejaku51 committed:
Revert "pAI TGUI conversion 1"

This reverts commit 1825abb8ddd0b8177ddef88f4c064d055d86be66.

Dejaku51 committed:
Add pAI internal GPS to possible software

Dejaku51 committed:
pAIs now use a radial menu to choose their holochassis skins

Dejaku51 committed:
Adds the ability for borgs to wear drones and pAIs as hats.

Dejaku51 committed:
Allows pAI's to print photographs for 10 ram

Dejaku51 committed:
Cleans up paicard and aicard code

Dejaku51 committed:
pAI's with Binary Encryption keys installed can use them

Dejaku51 committed:
Fixes oversight where pAIs are not able to select headset radio channels from encryption keys

Dejaku51 committed:
pAI minor refactor and cyborg PDA code/visual improvements

Dejaku51 committed:
pAI TGUI

Dejaku51 committed:
Removed bear and auto update

Dejaku51 committed:
Fixed pAI not being able to interact with its panel while in card

Dejaku51 committed:
Forgot about that when experimenting

Dejaku51 commented:
I'm not gonna wait a month for someone to review it or smth. Feel free to create that PR yourself

Dejaku51 commented:
Fuck it, im just gonna leave it open. If any maintainer gonna review it please ping me on discord (Dejaku51#9677) so i can make changes

Dejaku51 committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into pai_expansion

ike709 commented on code/controllers/subsystem/pai.dm:
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
+	to_chat(user, "<span class='notice'>You have requested pAI assistance.</span>")
+	var/mutable_appearance/alert_overlay = mutable_appearance('icons/obj/aicards.dmi', "pai")
+	notify_ghosts("[user] is requesting a pAI personality! Use the pAI button to submit yourself as one.", source=user, alert_overlay = alert_overlay, action=NOTIFY_ORBIT, header="pAI Request!", ignore_key = POLL_IGNORE_PAI)
+	addtimer(CALLBACK(src, .proc/request_again), 10 SECONDS)
```
Why isn't this a `VARSET_CALLBACK()`?

ike709 commented on code/controllers/subsystem/pai.dm:
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
+		if(!paicard.pai)
+			paicard.alertUpdate()
+	to_chat(usr, "<span class='notice'>Your pAI candidacy has been submitted!</span>")
+	addtimer(CALLBACK(src, .proc/submit_again), 10 SECONDS)
```
Why isn't this a `VARSET_CALLBACK()`?

ike709 commented on code/__DEFINES/pai.dm:
```diff
@@ -0,0 +1,2 @@
+/// Total percent required to hack a door.
+#define HACK_COMPLETE 100
```
Including a new file for one define is kind of bleh IMO.

ike709 commented on code/controllers/subsystem/pai.dm:
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
+	/// Prevents a crew member from hitting "request pAI"
+	var/request_spam = FALSE
+	/// Prevents a pAI from submitting itself repeatedly and sounding an alert.
+	var/submit_spam = FALSE
```
These are subsystem vars. Why does one person submitting a request once cause everyone to be ratelimited?

ike709 commented on code/controllers/subsystem/pai.dm:
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
-		candidate = new /datum/paiCandidate()
-		candidate.key = M.key
+/// Created when a user clicks the "pAI candidate" window
+/datum/pai_candidate
```
I'm not sure if these are filtered, but all of them (minus `key` & `ready`) need to be chat filtered when they're inputted.

ike709 commented on code/controllers/subsystem/pai.dm:
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
+		if(checked_candidate.key == user.key)
+			candidate = checked_candidate
+			return candidate
+	return null
```
Redundant.

ike709 commented on code/controllers/subsystem/pai.dm:
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
+	for(var/datum/pai_candidate/checked_candidate as anything in candidates)
+		if(checked_candidate.key == user.key)
+			candidate = checked_candidate
+			return candidate
```
Why not just make `candidates` an assoc list of key & datum?

ike709 commented on code/controllers/subsystem/pai.dm:
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
+	playsound(src, 'sound/machines/ping.ogg', 20, TRUE)
+	to_chat(user, "<span class='notice'>You have requested pAI assistance.</span>")
+	var/mutable_appearance/alert_overlay = mutable_appearance('icons/obj/aicards.dmi', "pai")
+	notify_ghosts("[user] is requesting a pAI personality! Use the pAI button to submit yourself as one.", source=user, alert_overlay = alert_overlay, action=NOTIFY_ORBIT, header="pAI Request!", ignore_key = POLL_IGNORE_PAI)
```
Does this check pAI bans?

ike709 commented on code/controllers/subsystem/pai.dm:
```diff
@@ -3,198 +3,168 @@ SUBSYSTEM_DEF(pai)
+	/// User inputted behavior description
+	var/description
+	/// User's ckey - not input
+	var/key
```
This comment is a lie, you're using their `key` everywhere when it really should be `ckey` everywhere except `IS_GUEST_KEY()`.

Dejaku51 replied:
so should i change `key` in comment to `ckey` or change every `key` in my code to `ckey` ?

ike709 replied:
Latter, minus the guest key check.

ike709 commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -1094,24 +1096,24 @@ GLOBAL_LIST_EMPTY(PDAs)
 	if(aicamera.stored.len)
-		var/add_photo = input(user,"Do you want to attach a photo?","Photo","No") as null|anything in list("Yes","No")
-		if(add_photo=="Yes")
+		var/add_photo = input(user,"Do you want to attach a photo?","Photo","No") as null|anything in list("Yes","No") /// Change to TGUI List once ported
```
No.

Archanial replied:
(ike meant comment, we're not porting tgui lists)

ike709 commented on code/game/objects/items/devices/gps.dm:
```diff
@@ -11,7 +11,10 @@
+	add_gps_component()
+
+/// Adds the GPS component to this item.
+/obj/item/gps/proc/add_gps_component()
```
```suggestion
/obj/item/gps/proc/add_gps_component()
	return
```

ike709 commented on code/game/objects/items/devices/paicard.dm:
```diff
@@ -16,156 +21,167 @@
+			for(var/datum/pai_candidate/checked_candidate as anything in SSpai.candidates)
+				if(params["key"] == checked_candidate.key)
+					candidate = checked_candidate
+					break
```
Look another place where it being assoc would simplify things.

Dejaku51 committed:
requested changes

Dejaku51 commented:
oh god there is more

ike709 commented on code/modules/mob/living/silicon/pai/pai.dm:
```diff
@@ -95,8 +111,33 @@
 	return FALSE
 
+/mob/living/silicon/pai/handle_atom_del(atom/A)
+	if(A == hacking_cable)
```
Couldn't this all be a switch()?

Dejaku51 replied:
don't think so, switch requires constant expression

ike709 reviewed (changes requested):
Gave this a once-over from ~2-6am whenever I was waiting on something to happen in Minecraft.

Looked at everything except the software file.

Also don't just manually resolve reviews without leaving a comment explaining why it's resolved.

Dejaku51 commented:
> Gave this a once-over from ~2-6am whenever I was waiting on something to happen in Minecraft.
> 
> Looked at everything except the software file.
> 
> Also don't just manually resolve reviews without leaving a comment explaining why it's resolved.

i just fixed these things that i marked as resolved

ike709 commented on code/__DEFINES/misc.dm:
```diff
@@ -467,3 +467,7 @@ GLOBAL_LIST_INIT(pda_styles, sortList(list(MONO, VT, ORBITRON, SHARE)))
+
+// pAI
+
+#define HACK_COMPLETE 100 // Total percent required to hack a door.
```
Don't make a define for 100?

ike709 commented:
Unaddressed reviews, fails CI

Dejaku51 committed:
Fix Conflicts

Dejaku51 committed:
forgot about comma

Dejaku51 committed:
fixes

Dejaku51 committed:
Remove comment

Dejaku51 committed:
key turned into ckey

Dejaku51 committed:
add return to add_gps_component

Dejaku51 committed:
deleted HACK_COMPLETE

Dejaku51 commented:
fuck you github actions

PowerfulBacon commented:
Dchat broke, could be related.
Runtime in paicard.dm, line 204: Cannot read [ckey].ready 
[09:29:45] Runtime in mobs.dm, line 434: Cannot read null.prefs
proc name: deadchat broadcast (/proc/deadchat_broadcast)

PowerfulBacon reviewed (changes requested):
Breaks GPSs, provided runtimes above and dchat no longer works (is last one related?)

PowerfulBacon commented:
PAI candidate loading doesn't work

Dejaku51 committed:
should be good now

Dejaku51 committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into pai_expansion

Dejaku51 committed:
fixes gps

Dejaku51 committed:
fixes once again

Dejaku51 committed:
bruh

Dejaku51 committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into pai_expansion

Dejaku51 committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into pai_expansion

Dejaku51 committed:
Loading candidate finally works...

jlsnow301 commented:
issue with candidate state is still present here, check this for fix
https://github.com/tgstation/tgstation/pull/65071

i am also saddened that you removed the pai's cat wallpaper

PowerfulBacon commented:
Fix candidate issue and this PR will be next in line for testmerge.

Dejaku51 commented:
> i am also saddened that you removed the pai's cat wallpaper

i liked it but removal was requested by people on discord

Dejaku51 commented:
> issue with candidate state is still present here, check this for fix
> https://github.com/tgstation/tgstation/pull/65071

i thought i did that huh

Dejaku51 committed:
Fixes tabs on pAI candidate menu

PowerfulBacon reviewed (changes requested):
pAI laws should be stated in the chat when they change.
Display image button doesn't work
ui_act should return true, the UI is very unresponsive when updating.
When in holoform folded out, I randomly take damage constantly.
![image](https://user-images.githubusercontent.com/26465327/170873690-a8d5c636-3e42-4c5f-a95d-91461858d102.png)
pAIs also have no way to leave the game themselves and other players cannot delete their pAIs now apparently.

PowerfulBacon reviewed (dismissed):
Suiciding as a pAI doesnt remove you from the list of pAIs that can be registered (Does it do this if you are downloaded already)

Dejaku51 committed:
fixes! + Wiping as pAI

Dejaku51 committed:
ok

Dejaku51 commented:
> Suiciding as a pAI doesnt remove you from the list of pAIs that can be registered (Does it do this if you are downloaded already)

Fixed

> pAI laws should be stated in the chat when they change.

Should work now

> Display image button doesn't work

Fixed

> ui_act should return true, the UI is very unresponsive when updating.

Done

> When in holoform folded out, I randomly take damage constantly.

Couldn't reproduce, probably something was happening during that particular round

> pAIs also have no way to leave the game themselves and other players cannot delete their pAIs now apparently.

Added wipe button

FortSumter commented:
I like this a lot. One thing I've wanted as a PAI was to vent crawl as a hologram. Mechanically this shouldn't be a problem as a PAI can't mess with anything on it's own bit it would be able to spy on people for antags or security while also not getting stuck behind doors when wandering.

I'd also like to be able to jump into people's hands back again. It was such a wholesome feature!

Dejaku51 committed:
Merge branch 'master' into pai_expansion

Dejaku51 committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into pai_expansion

Dejaku51 committed:
fixes etc

Dejaku51 committed:
pais are now stored in DB prefs!

Dejaku51 committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into pai_expansion

Dejaku51 committed:
fixes, not yet tested

itsmeow commented on code/modules/mob/living/silicon/silicon.dm:
```diff
@@ -24,7 +24,7 @@
 	hud_possible = list(ANTAG_HUD, DIAG_STAT_HUD, DIAG_HUD, DIAG_TRACK_HUD)
 
-	var/obj/item/radio/borg/radio = null //All silicons make use of this, with (p)AI's creating headsets
+	var/obj/item/radio/borg/pai/radio = null //All silicons make use of this, with (p)AI's creating headsets
```
This is for all silicons
you can assign a pAI radio to it but don't change the base type

Dejaku51 replied:
fixed

itsmeow commented on code/modules/mob/living/silicon/pai/software.dm:
```diff
@@ -206,7 +204,7 @@
 		return
 	to_chat(pai, "<span class='notice'>Requesting a DNA sample.</span>")
-	var/confirm = alert(master, "[pai] is requesting a DNA sample from you. Will you allow it to confirm your identity?", "Checking DNA", list("Yes", "No"))
+	var/confirm = input(master, "[pai] is requesting a DNA sample from you. Will you allow it to confirm your identity?", "Checking DNA") as anything in list("Yes", "No")
```
this will not work

Dejaku51 replied:
now?

itsmeow replied:
Looks correct

itsmeow commented on code/modules/mob/living/silicon/pai/software.dm:
```diff
@@ -4,24 +4,24 @@
+															"Atmosphere Sensor" = 5,
+															"Photography Module" = 5,
+															"Camera Zoom" = 10,
+															"Printer Module" = 10,
```
make all of the software names into DEFINEs please

Dejaku51 replied:
done

itsmeow commented on code/game/objects/items/devices/radio/radio.dm:
```diff
@@ -456,3 +456,9 @@
+
+/obj/item/radio/borg/pai
+	name = "pai radio"
+/obj/item/radio/borg/pai/ui_state(mob/user)
```
spacing between these two lines please

Dejaku51 replied:
removed entirely

itsmeow commented on code/game/objects/items/devices/paicard.dm:
```diff
@@ -132,6 +134,8 @@
 				pai.laws.set_zeroth_law("Serve your master.")
 				pai.emittersemicd = FALSE
 		if("set_laws")
+			if(!in_range(src, usr))
```
this needs to go AFTER the stripped_multiline_input

Dejaku51 replied:
done

itsmeow commented on code/game/objects/items/devices/paicard.dm:
```diff
@@ -14,6 +14,7 @@
 	///what emotion icon we have. handled in /mob/living/silicon/pai/Topic()
 	var/emotion_icon = "off"
 	resistance_flags = FIRE_PROOF | ACID_PROOF | INDESTRUCTIBLE
+	emag_toggleable = TRUE
```
don't do this, override can_emag and return true like the sleeper does

Dejaku51 replied:
done (at least i think i did it correctly)

itsmeow commented on tgui/packages/tgui/interfaces/PaiInterface.tsx:
```diff
@@ -71,36 +71,36 @@ seen as the ideal, unwavering human companion that you are.`;
-  'internal gps': 'A tool that allows you to track your location.',
-  'universal translator': 'Translation module for non-common languages.',
+  'Internal GPS': 'A tool that allows you to track your location.',
+  'Universal Translator': 'Translation module for non-common languages.',
```
I hate that this isn't a field on the software types passed in through static data

Dejaku51 replied:
is it a request or just complaining

itsmeow replied:
It is a request phrased as complaining

itsmeow replied:
It's also optional though

Dejaku51 committed:
fixes! i think?

Dejaku51 committed:
Merge branch 'master' into pai_expansion

Dejaku51 committed:
uhhh... yes?

Dejaku51 committed:
Merge branch 'pai_expansion' of https://github.com/Dejaku51/BeeStation-Hornet into pai_expansion

Dejaku51 committed:
pAI can interact with radio while inside card

Dejaku51 committed:
people didn't like it

itsmeow commented on code/modules/mob/living/silicon/pai/pai.dm:
```diff
@@ -375,7 +373,7 @@
 
 /obj/item/paicard/should_emag(mob/user)
-	return pai && ..()
+	return TRUE
```
return `!!pai`, my bad

Dejaku51 replied:
done

Dejaku51 committed:
emag

Dejaku51 committed:
pAI can now climb ladders

Dejaku51 committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into pai_expansion

Dejaku51 committed:
linter fix

itsmeow reviewed (approved):
1 year. Amazing.

ike709 reviewed (approved):
YOLO
